### PR TITLE
Documentation Documentation Documentation

### PR DESCRIPTION
**Pull request by MikeInnes**
Here's a first pass at writing some meta-documentation for the documentation system. Documentation.

I'll hold off on merging this until the type stuff actually, you know, works, and open it up for feedback in the meantime. Let me know if there's anything you're unsure about that this doesn't clear up.

### CODE REVIEW

File: doc/manual/documentation.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+.. doctest::
+
+    ?fft
+    ?@time
+    ?r""
+
+will bring up docs for the relevant function, macro or string macro respectively.
+In Juno using ``Ctrl-D`` will bring up documentation for the object under the
+cursor.
+
```
Comment (quinnj) [inline doc/manual/documentation.rst]: Is Juno mentioned anywhere else in the manual? I don't mind including it, but we may want to include more info about it and other relevant commands perhaps.

File: doc/manual/documentation.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+Documentation is interpreted as `Markdown <http://en.wikipedia.org/wiki/Markdown>`_,
+so you can use indentation and code fences to delimit code examples from text.
+
+.. doctest::
+
+    @doc doc"""
+      The `@bar` macro will probably make your code 2x faster or something. Use
+      it like this:
+
+          @bar by_drink_for("Mike")
```
Comment (simonbyrne) [inline doc/manual/documentation.rst]: by -> buy?

**Comment by timholy** [comment]
Is it possible to document whole modules? I'm thinking along the lines of
```
help?>Optim
# Displays the README.md for Optim, if nothing more specific has been defined
```

**Comment by one-more-minute** [comment]
Not yet, but that's a really good idea. Feel free to open an issue to remind me.

**Comment by timholy** [comment]
I was pleased to discover that this works:
```julia
for (f, op) in ((:add, :+), (:subtract, :-), (:multiply, :*), (:divide, :/))
    @eval begin
        $f(a,b) = $op(a,b)
    end
end
@doc "`add(a,b)` adds `a` and `b` together" ->add
@doc "`subtract(a,b)` subtracts `b` from `a`" ->subtract
```
Feel free to copy/paste this demo in.

In general, I'd say this PR only covers the surface of the system, and a somewhat more extensive treatment would be helpful. But this is a great start.

File: doc/manual/documentation.rst [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+.. doctest::
+
+    VERSION < v"0.4-" && using Docile
+
+    @doc doc"Tells you if there are too foo items in the array." ->
+    foo(xs::Array) = ...
+
+Documentation is interpreted as `Markdown <http://en.wikipedia.org/wiki/Markdown>`_,
+so you can use indentation and code fences to delimit code examples from text.
```
Comment (timholy) [inline doc/manual/documentation.rst]: On the latter point (automatically including the syntax), I see you've argued elsewhere that perhaps it's better not to do this. I can see your point.

**Comment by one-more-minute** [comment]
Glad to see people excited about this! I've tried to address the feedback/spelling mistakes etc. (apologies for taking so long on this) and in the interest of getting this out there I'll go ahead and merge. If I've missed any of the feedback here or anything else could be clearer, please do open an issue and ping me.

**Comment by timholy** [comment]
Yay! Thanks, @one-more-minute.

**Comment by ViralBShah** [comment]
Fantastic! Looking forward to everyone transitioning to the new docs now.

**Comment by hayd** [comment]
When should an effort begin to migrate Base? Do we first need to have the manual populated from `@doc` (so that documentation isn't duplicated) or is this already set up?

**Comment by ViralBShah** [comment]
That is the first step. We also then need to hook up `help()`, which is the easy part. I wonder if someone has a script that does the migration from the current format to the markdown format we want. We should probably also strive to structure each entry a bit better (R does a good job), and there was discussion on this in another issue, which I will dig up.

**Comment by jiahao** [comment]
So is there an accepted syntax for cross-referencing other functions/types in the new documentation format?

**Comment by hayd** [comment]
Currently works with REPL help mode and `Base.@help` but not with `Base.help` ?

Structure discussion was here: https://github.com/JuliaLang/julia/issues/8966. +1 for cross referencing/related functions as a standard section (as discussed in that issue).

This PR notes

>Documentation is very free-form; there are no set formatting restrictions or strict conventions. It's hoped that best practices will emerge fairly naturally as package developers learn to use the new system.

IME best practices rarely emerge naturally, would be great to come up with an example(s) of a "good" doc.

**Comment by jiahao** [comment]
@hayd I'm concerned that the migration from the current ReST format will be lossy, since the current docs are quite extensively cross-referenced.

Cross-references as a section like "See also:" is insufficient. Most of the current cross-references are inline.

**Comment by jiahao** [comment]
> IME best practices rarely emerge naturally, would be great to come up with an example(s) of a "good" doc.

Have you seen the help for functions like `eachindex`, `lufact`, `eigs` or `fft` recently? The Base docs already use nontrivial features of ReST for tables, executable code snippets, citations, cross-references and equations. None of these features exist in standard Markdown.

Not to denigrate the work put into the `@doc` macro so far, but we need support for all these features at the very least before even considering the migration of the standard library docs.

**Comment by ViralBShah** [comment]
These are all valid points. 

@MichaelHatherly @one-more-minute Should we track the various features we need before we can migrate as a separate issue?

**Comment by MichaelHatherly** [comment]
Yeah, I tried some time ago to convert some docs from rst to md using pandoc and it wasn't really a 1-1 conversion. @jiahao's definitely right. A new issue tracking needed features would probably be worth doing.

**Comment by one-more-minute** [comment]
I think it's completely fair to focus on package documentation for now – that's the much more urgent need, nice though it would be to have colourful docs for Base.

@jiahao is right that Base is a bit more complicated, though the hard part is mainly decisions like where documentation should live and whether we want to build to RST etc. We have #8967 open where we can discuss those decisions, if anyone has any opinions on them.

Things like references are pretty much trivial to implement, e.g.

```julia
type Reference
  obj
end

ref(x) = Reference(x)

writemime(io, "text/restructured", ::Reference) = # ... whatever the appropriate syntax is ...

doc"See also $(ref(fft))"
```
We could also just specialise this to
```julia
doc"See also [fft]"
```
but I'd need to think some more about how that plays nicely with links.

**Comment by hayd** [comment]
@one-more-minute I wonder if a nice syntax would be `doc"See also $[fft]"`.

Does this have to be something packages can define/overwrite - so that their references link to the correct page ?

**Comment by one-more-minute** [comment]
The `Reference` type itself could certainly be used by other packages, etc., but it would only make sense when outputting to formats like RST that support references directly. I'm not sure if you could do other formats like HTML generically.

**Comment by tshort** [comment]
Mkdocs supports cross references, tables, and Latex equations (with a plugin). References are normal markdown links. Here are two examples:
```julia
See [IdealDiode](../../lib/electrical/#idealdiode) and [Event](#event).
```
One references a function cited in another file, and one references a link in the same file. That might not be as smart as the Rst approach, but it gets the job done.

Mkdocs could be used for the online docs, and features could be incrementally added to Julia's markdown support to improve live HTML.

**Comment by hayd** [comment]
I wonder if this should/could be done with @MichaelHatherly's [Lexicon](https://github.com/MichaelHatherly/Lexicon.jl). It would be pretty awesome if the docs for the language were generated (from code) in exactly the same way as packages (with the same automated tool).

@jiahao I have a PR for MD tables, so that's part of the way there :). Executable code snippets could just be the same as Code objects, but labelled as such... Perhaps we need a tracking issue for all the features needed but we don't yet have?

*As discussed before, some kind of MetaDoc so you can pass a sequence of labelled sections would be really nice - this would also avoid inconsistent standards for headers in docstrings (`#` vs `##` vs `###`).*

**Comment by MichaelHatherly** [comment]
@hayd, I think it comes down to what should be in base and what shouldn't, ie. https://github.com/JuliaLang/julia/issues/5155. Interesting enough, with Docile's new plain docstrings (without `@doc`) it *could* be possible to have docstrings without adding anything to `Base`. Not sure how others feel about that though.

**Comment by hayd** [comment]
@MichaelHatherly IMO it doesn't need to be included in base to required for building the docs, I'm not advocating that!

**Comment by MichaelHatherly** [comment]
@hayd, thanks for clarifying that.

**Pull request merged**